server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -152,16 +185,35 @@ public RepositoryData addSnapshot(final SnapshotId snapshotId,
         Map<String, SnapshotState> newSnapshotStates = new HashMap<>(snapshotStates);
         newSnapshotStates.put(snapshotId.getUUID(), snapshotState);
         Map<IndexId, Set<SnapshotId>> allIndexSnapshots = new HashMap<>(indexSnapshots);
-        for (final IndexId indexId : snapshottedIndices) {
+        for (final IndexId indexId : shardGenerations.indices()) {
             allIndexSnapshots.computeIfAbsent(indexId, k -> new LinkedHashSet<>()).add(snapshotId);
         }
-        return new RepositoryData(genId, snapshots, newSnapshotStates, allIndexSnapshots);
+        return new RepositoryData(genId, snapshots, newSnapshotStates, allIndexSnapshots,
+                                  ShardGenerations.builder().putAll(this.shardGenerations).putAll(shardGenerations).build());
+    }
+
+    /**
+     * Create a new instance with the given generation and all other fields equal to this instance.
+     *
+     * @param newGeneration New Generation
+     * @return New instance
+     */
+    public RepositoryData withGenId(long newGeneration) {
```

Comment:
This came with https://github.com/elastic/elasticsearch/commit/4849c3e1fe15cbb343b007be09319de2eeef9430

Are the other changes also ported?

Reply:
All changes were already in, i took the related tests as well now.